storage: TestStoreRangeMergeRaftSnapshot failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/1cd7dcfe1bbf487a714c7f00a5dd709360dbe9e4

Parameters:
```
TAGS=
GOFLAGS=
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestStoreRangeMergeRaftSnapshot PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1049322&tab=buildLog

```
I181211 06:36:03.895686 37049 storage/client_merge_test.go:2865  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181211 06:36:04.895891 37049 storage/client_merge_test.go:2865  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181211 06:36:05.896115 37049 storage/client_merge_test.go:2865  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181211 06:36:06.896334 37049 storage/client_merge_test.go:2865  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181211 06:36:07.896595 37049 storage/client_merge_test.go:2865  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181211 06:36:08.896840 37049 storage/client_merge_test.go:2865  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181211 06:36:09.897122 37049 storage/client_merge_test.go:2865  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181211 06:36:10.897380 37049 storage/client_merge_test.go:2865  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181211 06:36:11.897567 37049 storage/client_merge_test.go:2865  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181211 06:36:12.897781 37049 storage/client_merge_test.go:2865  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181211 06:36:13.897991 37049 storage/client_merge_test.go:2865  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
W181211 06:36:14.551227 37380 storage/store_rebalancer.go:226  [s3,store-rebalancer] StorePool missing descriptor for local store
I181211 06:36:14.898281 37049 storage/client_merge_test.go:2865  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181211 06:36:15.898528 37049 storage/client_merge_test.go:2865  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181211 06:36:16.898815 37049 storage/client_merge_test.go:2865  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181211 06:36:17.898934 37049 storage/client_merge_test.go:2865  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181211 06:36:18.899136 37049 storage/client_merge_test.go:2865  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181211 06:36:19.899320 37049 storage/client_merge_test.go:2865  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181211 06:36:20.899530 37049 storage/client_merge_test.go:2865  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181211 06:36:21.899693 37049 storage/client_merge_test.go:2865  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181211 06:36:22.899883 37049 storage/client_merge_test.go:2865  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181211 06:36:23.900094 37049 storage/client_merge_test.go:2865  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
W181211 06:36:24.901462 37272 gossip/gossip.go:1499  [n2] no incoming or outgoing connections
W181211 06:36:24.901564 37399 gossip/gossip.go:1499  [n3] no incoming or outgoing connections
--- FAIL: TestStoreRangeMergeRaftSnapshot (57.87s)
	client_merge_test.go:2865: condition failed to evaluate within 45s: expected store2 to apply at least 1 additional raft snapshot
		goroutine 37049 [running]:
		runtime/debug.Stack(0xa7a358200, 0xc420838cc0, 0x2e9afa0)
			/usr/local/go/src/runtime/debug/stack.go:24 +0xa7
		github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon(0x2ef6120, 0xc4215dc0f0, 0xc420838c90)
			/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:49 +0x10d
		github.com/cockroachdb/cockroach/pkg/storage_test.TestStoreRangeMergeRaftSnapshot(0xc4215dc0f0)
			/go/src/github.com/cockroachdb/cockroach/pkg/storage/client_merge_test.go:2865 +0xd1c
		testing.tRunner(0xc4215dc0f0, 0x2a4ab50)
			/usr/local/go/src/testing/testing.go:777 +0xd0
		created by testing.(*T).Run
			/usr/local/go/src/testing/testing.go:824 +0x2e0

```

SHA: https://github.com/cockroachdb/cockroach/commits/59792cfdedad10828fcbf6328fe4195b6de5d8f0

Parameters:
```
TAGS=
GOFLAGS=
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestStoreRangeMergeRaftSnapshot PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1051236&tab=buildLog

```
I181212 06:41:03.645958 37058 storage/client_merge_test.go:2865  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181212 06:41:04.646164 37058 storage/client_merge_test.go:2865  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181212 06:41:05.646379 37058 storage/client_merge_test.go:2865  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181212 06:41:06.646586 37058 storage/client_merge_test.go:2865  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181212 06:41:07.646771 37058 storage/client_merge_test.go:2865  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181212 06:41:08.646969 37058 storage/client_merge_test.go:2865  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181212 06:41:09.647142 37058 storage/client_merge_test.go:2865  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181212 06:41:10.647480 37058 storage/client_merge_test.go:2865  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181212 06:41:11.647659 37058 storage/client_merge_test.go:2865  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181212 06:41:12.647850 37058 storage/client_merge_test.go:2865  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
W181212 06:41:12.654824 37170 storage/store_rebalancer.go:226  [s1,store-rebalancer] StorePool missing descriptor for local store
I181212 06:41:13.648078 37058 storage/client_merge_test.go:2865  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181212 06:41:14.648345 37058 storage/client_merge_test.go:2865  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181212 06:41:15.648667 37058 storage/client_merge_test.go:2865  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181212 06:41:16.648910 37058 storage/client_merge_test.go:2865  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181212 06:41:17.649186 37058 storage/client_merge_test.go:2865  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181212 06:41:18.649499 37058 storage/client_merge_test.go:2865  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181212 06:41:19.649800 37058 storage/client_merge_test.go:2865  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181212 06:41:20.650086 37058 storage/client_merge_test.go:2865  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
W181212 06:41:20.975462 37273 storage/store_rebalancer.go:226  [s2,store-rebalancer] StorePool missing descriptor for local store
I181212 06:41:21.650393 37058 storage/client_merge_test.go:2865  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
W181212 06:41:22.652469 37005 storage/raft_transport.go:583  while processing outgoing Raft queue to node 3: rpc error: code = Unavailable desc = transport is closing:
W181212 06:41:22.652561 37445 storage/raft_transport.go:583  while processing outgoing Raft queue to node 1: rpc error: code = Unavailable desc = transport is closing:
--- FAIL: TestStoreRangeMergeRaftSnapshot (57.79s)
	client_merge_test.go:2865: condition failed to evaluate within 45s: expected store2 to apply at least 1 additional raft snapshot
		goroutine 37058 [running]:
		runtime/debug.Stack(0xa7a358200, 0xc421ba60c0, 0x2ee6880)
			/usr/local/go/src/runtime/debug/stack.go:24 +0xa7
		github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon(0x2f43420, 0xc4201d9680, 0xc421ba6090)
			/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:49 +0x10d
		github.com/cockroachdb/cockroach/pkg/storage_test.TestStoreRangeMergeRaftSnapshot(0xc4201d9680)
			/go/src/github.com/cockroachdb/cockroach/pkg/storage/client_merge_test.go:2865 +0xd1c
		testing.tRunner(0xc4201d9680, 0x2a93a90)
			/usr/local/go/src/testing/testing.go:777 +0xd0
		created by testing.(*T).Run
			/usr/local/go/src/testing/testing.go:824 +0x2e0

```

SHA: https://github.com/cockroachdb/cockroach/commits/cc2bf701b5acea511860a62a85f9e48fc00c7628

Parameters:
```
TAGS=
GOFLAGS=
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestStoreRangeMergeRaftSnapshot PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1068566&tab=buildLog

```
I181224 06:29:24.817853 38059 storage/client_merge_test.go:2885  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181224 06:29:25.820085 38059 storage/client_merge_test.go:2885  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181224 06:29:26.820257 38059 storage/client_merge_test.go:2885  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181224 06:29:27.820434 38059 storage/client_merge_test.go:2885  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181224 06:29:28.820590 38059 storage/client_merge_test.go:2885  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181224 06:29:29.820779 38059 storage/client_merge_test.go:2885  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181224 06:29:30.820969 38059 storage/client_merge_test.go:2885  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181224 06:29:31.821159 38059 storage/client_merge_test.go:2885  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181224 06:29:32.821325 38059 storage/client_merge_test.go:2885  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181224 06:29:33.821529 38059 storage/client_merge_test.go:2885  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181224 06:29:34.821742 38059 storage/client_merge_test.go:2885  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181224 06:29:35.821941 38059 storage/client_merge_test.go:2885  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181224 06:29:36.822245 38059 storage/client_merge_test.go:2885  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181224 06:29:37.822494 38059 storage/client_merge_test.go:2885  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181224 06:29:38.822694 38059 storage/client_merge_test.go:2885  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181224 06:29:39.822943 38059 storage/client_merge_test.go:2885  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181224 06:29:40.823187 38059 storage/client_merge_test.go:2885  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181224 06:29:41.823440 38059 storage/client_merge_test.go:2885  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
W181224 06:29:42.595992 38429 storage/store_rebalancer.go:226  [s3,store-rebalancer] StorePool missing descriptor for local store
I181224 06:29:42.823652 38059 storage/client_merge_test.go:2885  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
W181224 06:29:42.989527 38294 storage/store_rebalancer.go:226  [s2,store-rebalancer] StorePool missing descriptor for local store
I181224 06:29:43.823869 38059 storage/client_merge_test.go:2885  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
I181224 06:29:44.824147 38059 storage/client_merge_test.go:2885  SucceedsSoon: expected store2 to apply at least 1 additional raft snapshot
--- FAIL: TestStoreRangeMergeRaftSnapshot (52.58s)
	client_merge_test.go:2885: condition failed to evaluate within 45s: expected store2 to apply at least 1 additional raft snapshot
		goroutine 38059 [running]:
		runtime/debug.Stack(0xa7a358200, 0xc42162b950, 0x2f214c0)
			/usr/local/go/src/runtime/debug/stack.go:24 +0xa7
		github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon(0x2f7ea20, 0xc421802000, 0xc42162b920)
			/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:49 +0x10d
		github.com/cockroachdb/cockroach/pkg/storage_test.TestStoreRangeMergeRaftSnapshot(0xc421802000)
			/go/src/github.com/cockroachdb/cockroach/pkg/storage/client_merge_test.go:2885 +0xd20
		testing.tRunner(0xc421802000, 0x2aca738)
			/usr/local/go/src/testing/testing.go:777 +0xd0
		created by testing.(*T).Run
			/usr/local/go/src/testing/testing.go:824 +0x2e0

```

--- a/pkg/storage/client_merge_test.go
+++ b/pkg/storage/client_merge_test.go
@@ -57,6 +57,7 @@ import (
 	"github.com/gogo/protobuf/proto"
 	"github.com/pkg/errors"
 	"github.com/stretchr/testify/assert"
+	"go.etcd.io/etcd/raft/raftpb"
 )
 
 func adminMergeArgs(key roachpb.Key) *roachpb.AdminMergeRequest {
@@ -3088,7 +3089,7 @@ func TestStoreRangeMergeRaftSnapshot(t *testing.T) {
 	}
 
 	// Truncate the logs of the LHS.
-	{
+	index := func() uint64 {
 		repl := store0.LookupReplica(roachpb.RKey("a"))
 		index, err := repl.GetLastIndex()
 		if err != nil {
@@ -3104,12 +3105,27 @@ func TestStoreRangeMergeRaftSnapshot(t *testing.T) {
 		if _, err := client.SendWrapped(ctx, mtc.distSenders[0], truncArgs); err != nil {
 			t.Fatal(err)
 		}
-	}
+		return index
+	}()
 
 	beforeRaftSnaps := store2.Metrics().RangeSnapshotsNormalApplied.Count()
 
 	// Restore Raft traffic to the LHS on store2.
-	mtc.transport.Listen(store2.Ident.StoreID, store2)
+	mtc.transport.Listen(store2.Ident.StoreID, &unreliableRaftHandler{
+		rangeID:            aRepl0.RangeID,
+		RaftMessageHandler: store2,
+		dropReq: func(req *storage.RaftMessageRequest) bool {
+			// Make sure that even going forward no MsgApp for what we just
+			// truncated can make it through. The Raft transport is asynchronous
+			// so this is necessary to make the test pass reliably - otherwise
+			// the follower on store2 may catch up without needing a snapshot,
+			// tripping up the test.
+			//
+			// NB: the Index on the message is the log index that _precedes_ any of the
+			// entries in the MsgApp, so filter where msg.Index < index, not <= index.
+			return req.Message.Type == raftpb.MsgApp && req.Message.Index < index
+		},
+	})
 
 	// Wait for all replicas to catch up to the same point. Because we truncated
 	// the log while store2 was unavailable, this will require a Raft snapshot.

